src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -139,6 +139,121 @@ ASR::Module_t* load_module(Allocator &al, SymbolTable *symtab,
+                } else {
+                    return false;
+                }
+                break;
+            }
+            case (ASR::ttypeType::Complex) : {
+                ASR::Complex_t *a2 = ASR::down_cast<ASR::Complex_t>(&a);
+                ASR::Complex_t *b2 = ASR::down_cast<ASR::Complex_t>(&b);
+                if (a2->m_kind == b2->m_kind) {
+                    return true;
+                } else {
+                    return false;
+                }
+                break;
+            }
+            case (ASR::ttypeType::Logical) : {
+                ASR::Logical_t *a2 = ASR::down_cast<ASR::Logical_t>(&a);
+                ASR::Logical_t *b2 = ASR::down_cast<ASR::Logical_t>(&b);
+                if (a2->m_kind == b2->m_kind) {
+                    return true;
+                } else {
+                    return false;
+                }
+                break;
+            }
+            case (ASR::ttypeType::Character) : {
+                ASR::Character_t *a2 = ASR::down_cast<ASR::Character_t>(&a);
+                ASR::Character_t *b2 = ASR::down_cast<ASR::Character_t>(&b);
+                if (a2->m_kind == b2->m_kind) {
+                    return true;
+                } else {
+                    return false;
+                }
+                break;
+            }
+            default : return false;
+        }
+    }
+    return false;
+}
+
+template <typename T>
+bool argument_types_match(const Vec<ASR::ttype_t*> &args,
+        const T &sub) {
+    if (args.size() <= sub.n_args) {
+        size_t i;
+        for (i = 0; i < args.size(); i++) {
+            ASR::Variable_t *v = LFortran::ASRUtils::EXPR2VAR(sub.m_args[i]);
+            ASR::ttype_t *arg1 = args[i];
+            ASR::ttype_t *arg2 = v->m_type;
+            if (!types_equal(*arg1, *arg2)) {
+                return false;
+            }
+        }
+        for( ; i < sub.n_args; i++ ) {
+            ASR::Variable_t *v = LFortran::ASRUtils::EXPR2VAR(sub.m_args[i]);
+            if( v->m_presence != ASR::presenceType::Optional ) {
+                return false;
+            }
+        }
+        return true;
+    } else {
+        return false;
+    }
+}
+
+bool select_func_subrout(const ASR::symbol_t* proc, const Vec<ASR::ttype_t*> &args,
+                         const Location& loc, const std::function<void (const std::string &, const Location &)> err) {
+    bool result = false;
+    if (ASR::is_a<ASR::Subroutine_t>(*proc)) {
+        ASR::Subroutine_t *sub
+            = ASR::down_cast<ASR::Subroutine_t>(proc);
+        if (argument_types_match(args, *sub)) {
+            result = true;
+        }
+    } else if (ASR::is_a<ASR::Function_t>(*proc)) {
+        ASR::Function_t *fn
+            = ASR::down_cast<ASR::Function_t>(proc);
+        if (argument_types_match(args, *fn)) {
+            result = true;
+        }
+    } else {
+        err("Only Subroutine and Function supported in generic procedure", loc);
+    }
+    return result;
+}
```

Comment:
We should abstract this out from LFortran and LPython and put into `libasr` to be used by both.

Reply:
Yeah, it's just the modified version of https://github.com/lcompilers/libasr/blob/82d115cf2e0886e794c8ebe22ff04c094370acc0/src/libasr/asr_utils.cpp#L630. It is present in `libasr`, but isn't included in any namespace.

Reply:
What modifications were needed? Let's improve the function from `libasr` and use it instead.

In order to reduce complexity of the frontend, we need to try to figure out how to "offload" as much as functionality into libasr, where it is shared across multiple frontends (LPython and LFortran for now, more in the future).

Effectively, these are all auxiliary routines to make building ASR easier. In this case, building GenericProcedures and using them easier.

Reply:
I think they are almost identical, so I would use this from the `libasr` directly.